src/katgpucbf/xbgpu/engine.py

```diff
@@ -259,9 +273,171 @@ def capture_enable(self, enable: bool = True) -> None:
+            self._tx_free_item_queue.put_nowait(tx_item)
+
+        # TODO: The way this is imported will change once the OperationSequence is in place
+        self.send_stream = bsend.BSend(
+            output,
+            engine.heaps_per_fengine_per_chunk,
+            self.n_tx_items,
+            n_channels_per_substream=engine.n_channels_per_substream,
+            spectra_per_heap=engine.src_layout.n_spectra_per_heap,
+            timestamp_step=engine.rx_heap_timestamp_step,
+            send_rate_factor=engine.send_rate_factor,
+            channel_offset=engine.channel_offset_value,
+            context=context,
+            packet_payload=engine.dst_packet_payload,
+            stream_factory=lambda stream_config, buffers: make_bstream(
+                endpoints=[output.dst],
+                interface=engine.dst_interface,
+                ttl=engine.dst_ttl,
+                use_ibv=engine.dst_ibv,
+                affinity=engine.dst_affinity,
+                comp_vector=engine.dst_comp_vector,
+                stream_config=stream_config,
+                buffers=buffers,
+            ),
+            tx_enabled=init_tx_enabled,
+        )
+
+        self._populate_sensors()
+
+    def _populate_sensors(self) -> None:
+        sensors = self.engine.sensors
+        # Dynamic sensors
+        # TODO: Update once multiple beams are supported
+        sensors.add(
+            aiokatcp.Sensor(
+                int,
+                f"{self.output.name}-beng-clip-cnt",
+                "Number of complex samples that saturated.",
+                default=0,
+                initial_status=aiokatcp.Sensor.Status.NOMINAL,
+            )
+        )
+
+    def capture_enable(self, enable: bool = True) -> None:  # noqa: D102
+        # NOTE: Defaulting to handle its single output for now.
+        # Actual implementation will require some kind of 'stream_id' arg,
+        # which currently contravenes the base function signature.
+        self.send_stream.enable_substream(enable=enable)
+
+    async def gpu_proc_loop(self) -> None:  # noqa: D102
+        while True:
+            # Get item from the receiver loop.
+            # - Wait for the HtoD transfers to complete, then
+            # - Give the chunk back to the receiver for reuse.
+            rx_item = await self._rx_item_queue.get()
+            if rx_item is None:
+                break
+            await rx_item.async_wait_for_events()
+
+            tx_item = await self._tx_free_item_queue.get()
+            await tx_item.async_wait_for_events()
+            # TODO: Is it fine to update the timestamp like this?
+            tx_item.reset(rx_item.timestamp)
+
+            # Bind input buffers
+
+            # Queue GPU work
+            # - For now, just fill it with zeros
+            tx_item.buffer_device.zero(self._proc_command_queue)
+            tx_item.saturated.zero(self._proc_command_queue)
+
+            # Bind output buffers
```

Comment:
That should be before the GPU work.

Reply:
Related to the self-explanatory comments, I'd left this kind of comment in here mostly as placeholders/reminders. I don't think they actually need to persist beyond this PR - I'm sure we'll be able to figure out what goes where/order of operations later on.